### ISSUE TITLE
Expected dataloader behaviour

### ISSUE THREAD

**Issue by francisr** [lhotse-speech/lhotse#1509]
Hi,

I have a dataset of a few thousand hours, split into 2400 `recording.*.tar` and `cuts.*.jsonl.gz` files, stored on NFS, and the audio is encoded as FLAC.  
I'm trying to use Lhotse to replace my dataloader that fetches random segments on NFS, which can get slow occasionally.  

However I'm having a few issues:
- Calling `iter(loader)` can take more than 5 minutes.
- I get a throughput of about 4000 (so 4000 seconds per second) with 16 works. Of course it's going to depend on hardware, but I expected more a priori.
- There outliers in batch durations. I set a max of 1000s, and sometimes I get batches that have in the 800s range. Why does this happen?

Any comments?

This is the code I'm using to debug that.
```
import argparse
import glob
import logging
import math
import os
import time
from pathlib import Path
from typing import Iterable, List

import torch
from torch.utils.data import DataLoader

from lhotse import CutSet
from lhotse.dataset import DynamicBucketingSampler
from lhotse.dataset.dataloading import worker_init_fn


def parse_args() -> argparse.Namespace:
    ap = argparse.ArgumentParser()
    ap.add_argument("--shar", required=True, help="Shar glob")
    ap.add_argument("--effective-batch-seconds", type=float, required=True,
                    help="Target *global* batch size in seconds across all ranks.")
    ap.add_argument("--buckets", type=int, default=30, help="Num duration buckets.")
    ap.add_argument("--shuffle", action="store_true", help="Shuffle cuts & shards.")
    ap.add_argument("--seed", type=int, default=0, help="Seed for shuffling.")
    ap.add_argument("--num-workers", type=int, default=4, help="PyTorch DataLoader workers.")
    ap.add_argument("--max-steps", type=int, default=10, help="Stop after this many steps (per rank).")
    ap.add_argument("--print-n", type=int, default=5, help="Print first N cut IDs in each batch.")
    ap.add_argument("--dry-audio", action="store_true",
                    help="Do NOT touch audio/features; just iterate CutSets.")
    # DDP / manual override
    ap.add_argument("--world-size", type=int, default=None, help="Override WORLD_SIZE env.")
    ap.add_argument("--rank", type=int, default=None, help="Override RANK env.")
    return ap.parse_args()


def get_dist_info(args: argparse.Namespace) -> tuple[int, int]:
    # Read from torchrun env if available, else from CLI, else single-process.
    ws = int(os.environ.get("WORLD_SIZE", "1"))
    rk = int(os.environ.get("RANK", "0"))
    if args.world_size is not None:
        ws = args.world_size
    if args.rank is not None:
        rk = args.rank
    return ws, rk


class DebugDataset:
    """
    Minimal dataset that accepts a CutSet batch (from Lhotse sampler) and
    returns a small dict of debug stats. If --dry-audio is False, it will
    touch cut.load_audio() for 1 cut to verify payloads (but not store them).
    """
    def __init__(self, dry_audio: bool = True):
        self.dry = dry_audio

    def __getitem__(self, batch_cuts: CutSet):
        # batch_cuts is a CutSet (already the batch)
        cuts = list(batch_cuts)
        dur_sum = sum(c.duration for c in cuts)
        ids = [c.id for c in cuts]
        maybe_shape = None
        wavs = []
        if not self.dry and len(cuts) > 0:
            # Touch one sample to make sure shard payloads are readable.
            # This will actually decode audio for that cut.
            for cut in cuts:
                wav = cut.load_audio()
                wavs.append(wav)
                maybe_shape = tuple(wav.shape)  # (channels, samples)

        return {
            "num_cuts": len(cuts),
            "dur_seconds": dur_sum,
            "ids": ids,
            "probe_shape": maybe_shape,
            "audio": wavs,
        }

    # Required by DataLoader when batch_size=None and sampler yields CutSets:
    def __len__(self):
        # Unused in this debug script; DataLoader+sampler will handle iteration.
        return 2**31 - 1



def main():
    args = parse_args()
    world_size, rank = get_dist_info(args)

    # print("args.shar", args.shar)
    part_dirs = sorted(f for pat in args.shar for f in glob.glob(pat, recursive=True))
    # print("part_dirs", part_dirs)

    cuts_files = [
                str(p)
                for d in part_dirs
                for p in glob.glob(str(Path(d) / "cuts.*.jsonl.gz"))
            ]
    rec_files = [
                str(p)
                for d in part_dirs
                for p in glob.glob(str(Path(d) / "recording.*.tar"))
            ]
    cuts = CutSet.from_shar(
        fields={
            "cuts": cuts_files,
            "recording": rec_files,
        },
        split_for_dataloading=True,
        shuffle_shards=True,
        seed=0,
    )
    print(cuts)

    # Translate global effective batch seconds into per-rank target.
    per_rank_max_dur = args.effective_batch_seconds / max(1, world_size)

    sampler = DynamicBucketingSampler(
        cuts,
        buffer_size=2000000,
        max_duration=per_rank_max_dur,
        num_buckets=args.buckets,
        shuffle=False,
        drop_last=False,
        world_size=world_size,
        rank=rank,
        sync_buckets=False,
    )

    dset = DebugDataset(dry_audio=args.dry_audio)

    # DataLoader: batch_size=None because sampler yields full CutSets as batches.
    loader = DataLoader(
        dset,
        batch_size=None,
        sampler=sampler,
        num_workers=args.num_workers,
        worker_init_fn=worker_init_fn,
        persistent_workers=args.num_workers > 0,
    )

    print(f"[rank {rank}/{world_size}] starting debug loop; "
          f"per-rank max_dur={per_rank_max_dur:.2f}s; workers={args.num_workers}")

    t0 = time.time()
    iterloader = iter(loader)
    dt = time.time() - t0
    print(f"iter(loader) in {dt:.2f}s")

    t0 = time.time()
    batch = next(iterloader)
    dt = time.time() - t0
    print(f"First batch in {dt:.2f}s")

    t0 = time.time()
    tot_dur = 0
    for step, batch in enumerate(iterloader, 1):
        # batch is our dict from DebugDataset.__getitem__
        tot_dur += batch['dur_seconds']
        print(
            f"[rank {rank}] step {step:04d} | "
            f"cuts={batch['num_cuts']:4d} | "
            f"dur_s={batch['dur_seconds']:.2f} | "
            f"probe={str(batch['probe_shape']) if batch['probe_shape'] else ''}"
        )
        if step >= args.max_steps:
            break
    dt = time.time() - t0
    print(f"[rank {rank}] done {min(step, args.max_steps)} steps in {dt:.2f}s")

    throughput = tot_dur / dt
    print(f"Throughput: {throughput}")

    print(sampler.duration_bins)


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)  # or DEBUG
    main()

```

**Comment by nd-sketch** [comment]
My guess is that your buffer size is a bit large around 2M so it will take a while to pre-fill it, reducing it can lead to faster initialization. Regarding your problem of getting batches around 800s even if you set it to 1000s can be a cause of exceeding the max duration after fetching the next cut or segment. You can try setting `drop_last=True`

Reference: [Fallback when exceeding duration](http://github.com/lhotse-speech/lhotse/blob/master/lhotse/dataset/sampling/dynamic.py#L307-L320)

**Comment by pzelasko** [comment]
>        buffer_size=2000000,
You have an excessive buffer size, it will take very long to fill. Use 10-20k at most.

> I get a throughput of about 4000 (so 4000 seconds per second) with 16 workers. Of course it's going to depend on hardware, but I expected more a priori.

What's the throughput with 1 worker? Does it scale linearly?

> There outliers in batch durations. I set a max of 1000s, and sometimes I get batches that have in the 800s range. Why does this happen?

If you have an utterance with eg 210s duration, it would sample 4 examples leading to ~840s cumulative duration (including padding) and stop there.
